pipelines/sql_database/sql_database.py

```diff
@@ -1,121 +1,43 @@
-    current_val = cursor_state['last_value']
-    if current_val != value:
-        cursor_state['last_value'] = value
-        cursor_state['loaded_ids'] = []
-    if loaded_id is not None:
-        cursor_state['loaded_ids'].append(loaded_id)
-
-
-def _make_query(table: Table, cursor_column: Optional[str], unique_column: Optional[str]) -> Tuple[Select[Any], Optional[Column[Any]], Optional[Column[Any]]]:
-    """Make the SQL query object to select from table.
-    Returns a tuple with the query and optionally the cursor column as `sqlalchemy.Column` object
-    """
-    query = table.select()
-    if not cursor_column:
-        return query, None, None
-    cursor_col = table.c.get(cursor_column)
-    if cursor_col is None:
-        # Cursor column doesn't exist in table
-        # TODO: Should this print a warning or exception?
-        return query, None, None
-    query = query.order_by(cursor_col)
-    cursor_state = _get_cursor_state(table.name, cursor_column)
-    last_value = cursor_state['last_value']
-    loaded_ids = cursor_state['loaded_ids']
-    unique_col = table.c.get(unique_column)
-    if last_value is None:
-        return query, cursor_col, unique_col
-    query = query.where(cursor_col >= last_value)
-    if not not loaded_ids and unique_col is not None:
-        query = query.where(tuple_(cursor_col, unique_col).notin_(
-            [(last_value, lid) for lid in loaded_ids]
-        ))
-    return query, cursor_col, unique_col
-
-
-def table_rows(
-    engine: Engine,
-    table: Table,
-    cursor_column: Optional[str] = None,
-    unique_column: Optional[str] = None
-) -> Iterator[Dict[str, Any]]:
-    """Yields rows from the given database table.
-
-    :param engine: An `sqlalchemy.engine.Engine` instance configured for the database
-    :param table: The table to load data from
-    :param cursor_column: Optional column name to use as cursor for resumeable loading
-    :param unique_column: Optional column that uniquely identifies a row in the table for resumeable loading
-    """
-    query, cursor, unique = _make_query(table, cursor_column, unique_column)
-
-    conn: Connection
-    with engine.connect() as conn:
-        result = conn.execution_options(yield_per=1000).execute(query)
-        for partition in result.partitions():
-            for row in partition:
-                if cursor is not None:
-                    _set_last_value(
-                        table.name, cursor_column, row._mapping[cursor], row._mapping.get(unique.name) if unique is not None else None
-                    )
-                yield dict(row._mapping)
+from pipelines.sql_database.util import table_rows
 
 
 @dlt.source
 def sql_database(
-    database_url: str = dlt.secrets.value,
+    credentials: ConnectionStringCredentials = dlt.secrets.value,
     schema: Optional[str] = dlt.config.value,
     table_names: Optional[List[str]] = dlt.config.value,
     write_disposition: str = 'replace',
```

Comment:
this is still not being used. so in fact all the resources are "append", (or maybe I'm wrong?)
btw. if this is set to replace and the cursor column is specified then the past data will be deleted and only the increment will be kept. there may be use cases for that...

Reply:
This argument was left over from when I was experimenting initially. I removed it for now. Doesn't really make sense to have it there since you can pass it to `pipeline.run` anyway, right?